### PR TITLE
ID Site

### LINKED ISSUE #73: IDsite
We have in the framework spec to kick off the IDSite thing, but not vice versa. Should we add this into the spec for 1.0? Looks like express has a `/idSiteResult` endpoint to take the `StormpathToken`

### PR DESCRIPTION
**Pull request by edjiang**
Fixes #73. Documents what we currently do.

The only potentially controversial change is that I removed the `web.idSite.nextUri` configuration setting. This is because ID Site only works if `/register`, `/login`, etc endpoints are enabled in the configuration. As a result I think it makes more sense that it would respect the `nextUri` settings of those parts of the configuration.

### CODE REVIEW

**Comment by nbarbettini** [comment]
Awesome, thanks for compiling this. Looking forward to discussion.

File: idsite.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# ID Site
+
+Framework integrations must implement the `web.idSite.enabled` behavior.
```
Comment (robertjd) [inline idsite.md]: `web.idSite.enabled` -> `stormpath.web.idSite.enabled`

File: idsite.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# ID Site
+
+Framework integrations must implement the `web.idSite.enabled` behavior. 
+
+ID Site is used as a replacement for the `/login`, `/logout`, `/forgot`, and `/register` pages, which will redirect to ID Site instead of performing their default behavior.
```
Comment (robertjd) [inline idsite.md]: ID Site is a hosted login & registration application that simplifies single-sign on (SSO) flows by providing a common place for redirects and cookie storage.  If the developer has chosen to use ID Site, they likely have multiple Stormpath Applications that are redirecting the user to ID Site for authentication.  When the user authenticates at ID Site, they are returned to the original application (service provider, AKA the SP) with Stormpath Assertion JWT.  The application can verify this token and resolve the account that has authenticated.

If the developer needs to use ID Site, they will enable it with `stormpath.web.idSite.enabled`.  When this value is true, we should redirect the user to ID site when the following URIs are accessed with a GET request:
- `stormpath.web.login.uri`
- `stormpath.web.logout.uri`
- `stormpath.web.register.uri`
- `stormpath.web.forgot.uri`

**Comment by robertjd** [comment]
Thanks!  This is a good start, I added a comment that helps to clarify what ID Site is for. Note: there is more context that needs to go into the specification when we define how multi-tenancy will work with the integrations (the framework will need to feed some organization context to ID Site).

**Comment by edjiang** [comment]
Updated!

**Comment by edjiang** [comment]
One point: in the case of errors, I could redirect to the /login page, but if there's actually an error with something, we could get into a redirect loop. Any ideas for how we want to deal with this?

File: idsite.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+A new user was created. Follow the standard post-registration logic (like redirecting to the login page or autologin). 
+
+**AUTHENTICATED**
+
+A user logged in. Follow the standard login logic (like setting cookies, and redirecting the user)
+
+**LOGOUT**
+
+A user logged out. Follow the standard logout logic.
```
Comment (robertjd) [inline idsite.md]: Need a link here to what the logout logic is.

File: idsite.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+The `jwtResponse` from ID Site can have three states, with the following response logic: 
+
+**REGISTERED**
+
+A new user was created. Follow the standard post-registration logic (like redirecting to the login page or autologin). 
+
+**AUTHENTICATED**
+
+A user logged in. Follow the standard login logic (like setting cookies, and redirecting the user)
```
Comment (robertjd) [inline idsite.md]: Link to specifics.

File: idsite.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+`jwtResponse` - a [Stormpath JWT](http://docs.stormpath.com/guides/using-id-site/#handling-the-callback-to-your-application-from-id-site). This should be handled using the underlying SDK's Stormpath Token authenticator. 
+
+## Response
+
+The `jwtResponse` from ID Site can have three states, with the following response logic: 
+
+**REGISTERED**
+
+A new user was created. Follow the standard post-registration logic (like redirecting to the login page or autologin).
```
Comment (robertjd) [inline idsite.md]: Link all the things :)

File: idsite.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ID Site is a hosted login & registration application that simplifies single-sign on (SSO) flows by providing a common place for redirects and cookie storage. If the developer has chosen to use ID Site, they likely have multiple Stormpath Applications that are redirecting the user to ID Site for authentication. When the user authenticates at ID Site, they are returned to the original application (service provider, AKA the SP) with Stormpath Assertion JWT. The application can verify this token and resolve the account that has authenticated.
+
+If the developer needs to use ID Site, they will enable it with `stormpath.web.idSite.enabled`. When this value is true, we should redirect the user to ID site when the following URIs are accessed with a GET request:
+
+* `stormpath.web.login.uri`
+* `stormpath.web.logout.uri`
+* `stormpath.web.register.uri`
+* `stormpath.web.forgot.uri`
+
+In addition, enabling ID Site in the configuration enables an `/idSiteResult` endpoint.
```
Comment (robertjd) [inline idsite.md]: The callback URI is configurable and should be mentioned, `stormpath.web.idSite.uri`

**Comment by robertjd** [comment]
Redirecting to the /login page won't make sense, for ID Site callback errors, because the login page is disabled (should not be used) when ID Site is being used.  I think the current specification of a 401 Unauthorized view is okay, IF the framework logs the actual error - to help developers debug ID Site problems.  In express we just send the JSON error to the end-user, as this is has proven useful for debugging.

**Comment by edjiang** [comment]
Cool. I'm not comfortable with specifying that JSON be returned to the end user, as a potential reason someone might get the 401 Unauthorized is if they're trying to figure out how the `/idSiteResult` endpoint works. Do you think logging should be in the spec? I think we should just let the framework author decide at this moment.

**Comment by edjiang** [comment]
Updated

**Comment by omgitstom** [comment]
@edjiang @robertjd 

`idSiteResult`  needs to be called: TBA

**Comment by omgitstom** [comment]
@edjiang @robertjd

Clarify that ID Site overrides the routes

**Comment by edjiang** [comment]
At the last meeting, we decided to rename the `/idSiteResult` endpoint to something more generic, because this endpoint will be used not just for ID Site, but also for SAML callbacks, as well as social login once Stormpath takes over the auth code flow. 

After some thinking, I decided to rename the `/idSiteResult` endpoint to `/callbacks/stormpath`. I know this isn't ideal, as some people at the meeting explicitly discouraged using the `/callbacks/` directory as this will only be one endpoint once Stormpath's API takes over social login.

However, I believe this is the right decision for 1.0 because we will need to support the social callbacks for the lifetime of 1.x, and this will keep the naming conventions consistent. 

Our configuration currently looks like the following: 

``` yaml
    idSite:
      enabled: false
      uri: "/idSiteResult"
      loginUri: ""
      forgotUri: "/#/forgot"
      registerUri: "/#/register"
    social:
      facebook:
        uri: "/callbacks/facebook"
        scope: "email"
      github:
        uri: "/callbacks/github"
        scope: "user:email"
      google:
        uri: "/callbacks/google"
        scope: "email profile"
      linkedin:
        uri: "/callbacks/linkedin"
        scope: "r_basicprofile r_emailaddress"
```

Thoughts?

**Pull request merged**